libs/traits/src/rewards.rs

```diff
@@ -182,6 +182,24 @@ pub trait CurrencyGroupChange {
 	fn currency_group(currency_id: Self::CurrencyId) -> Option<Self::GroupId>;
 }
 
+pub trait RewardIssuance {
+	/// Type used to identify the beneficiary.
+	type AccountId;
+
+	/// Type used to identify the currency
+	type CurrencyId;
+
+	/// Type used as balance for all currencies and reward.
+	type Balance;
```

Comment:
Seeing the implementations of these traits that end up using generic for specifying the types, do we really need the associated types in the trait? Why not `RewardIssuance<AccountId, CurrencyId, Balance`>?

Reply:
I am not a big fan of using generics. It forces consumers to define them while associated types have the advantage of being "unknonw".

Reply:
I suppose @lemunozm 's point is that the [two implementations](https://github.com/centrifuge/centrifuge-chain/pull/1209/files#diff-b8fa994533f9de709bf0781bb706a698f7afdadda84d688a82c48823227d7f0cR25-R26) of the trait use generics, so why not bubble that down? In the current version, I tried to continue the existing style.

Reply:
I know, but it is no advantage to use generics.

Reply:
My rule of thumb for this is:
- If the caller chose the types => generics
- If the implementer can chose the types => associated types

It's true that in Substrate most of the traits are implemented by the pallet that has the `Config` trait, and the associated types can be linked to any `T::Whatever`, so the implementer most of the time knows which types choose for the associated types.

In this case, because the implementer is not the pallet, it needs to expose them anyways as a generic.

Reply:
Can't the implementer always choose the type?

The caller can always enforce the type via `TraitA<Associated1=...>`. My personal opinion is just that generics are less flexible while providing no real advantage. Unless you wanna allow multiple implementations of this trait for the same implementer.

Reply:
Shouldn't have started this. ^^ I actually don't have strong such strong feelings about that, so do what works best for you.

Reply:
I propose to keep as is.